### PR TITLE
Task2 Signup & Login Page

### PR DESCRIPTION
**Pull request by nitzanpap**
Task2: Added the Signup & Login Page

### CODE REVIEW

File: lib/widgets/input_field_widgets/email_input_field.dart [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return AppInputField(
+      labelText: 'Your email',
+      keyboardType: TextInputType.emailAddress,
+      onChanged: onChanged,
+      textInputAction: TextInputAction.next,
+      validator: Validations.emailValidator,
+      hintText: 'john@gmail.com',
```
Comment (abenartz) [inline lib/widgets/input_field_widgets/email_input_field.dart:25]: We don't use some demo email for hint just use 'Email' as the hint

File: lib/widgets/input_field_widgets/password_input_field.dart [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  const PasswordInputField.reEnteredPassword({
+    super.key,
+    required this.onChanged,
+    this.textInputAction,
+    required this.originalPassword,
+    this.isReEnterPasswordInput = true,
+    this.labelText = 'Re-Enter Password',
+  });
+
+  String? reEnteredPasswordValidator(reEnteredPassword) {
```
Comment (abenartz) [inline lib/widgets/input_field_widgets/password_input_field.dart:34]: Name doesn't reveal what the func is doing and what return

File: lib/widgets/signup_login_form.dart [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    updateProvider(
+        provider: provider, formEmail: formEmail, formPassword: formPassword);
+    // TODO: Implement here, authenticating with firebase and moving to the next page.
+    print('Valid Form!');
+    return;
+  }
+
+  bool isFormValid(formKey) => formKey.currentState!.validate();
+
+  void updateProvider({
```
Comment (abenartz) [inline lib/widgets/signup_login_form.dart:159]: This function should be in the provider and not in the widget.
You can create inside the provider function called logister(email, password).
Inside the provider we will pass the params to firebase.

Next lesson we will talk about our architecture design.

**Review by abenartz — CHANGES_REQUESTED** [review]

File: lib/widgets/signup_login_form.dart (hunk 3 above)
Comment (nitzanpap) [reply lib/widgets/signup_login_form.dart]: @abenartz Done.

File: lib/widgets/signup_login_form.dart [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  }) {
+    if (!isFormValid(formKey)) {
+      print('Invalid input somewhere in the form!');
+      return;
+    }
+    updateProvider(
+        provider: provider, formEmail: formEmail, formPassword: formPassword);
+    // TODO: Implement here, authenticating with firebase and moving to the next page.
+    print('Valid Form!');
+    return;
```
[the remark this answers is not in the document]
Comment (nitzanpap) [reply lib/widgets/signup_login_form.dart]: @abenartz Done.

File: lib/widgets/input_field_widgets/password_input_field.dart (hunk 2 above)
Comment (nitzanpap) [reply lib/widgets/input_field_widgets/password_input_field.dart]: @abenartz Done.

File: lib/widgets/input_field_widgets/email_input_field.dart (hunk 1 above)
Comment (nitzanpap) [reply lib/widgets/input_field_widgets/email_input_field.dart]: @abenartz Done.

File: lib/widgets/button_widgets/primary_button.dart [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import 'package:dispatcher/widgets/button_widgets/base_button.dart';
+
+class PrimaryButton extends BaseButton {
+  // final Color color;
```
[the remark this answers is not in the document]
Comment (nitzanpap) [reply lib/widgets/button_widgets/primary_button.dart]: @abenartz Done.

File: lib/helper_functions/validations.dart [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    if (!hasDigitRegEx.hasMatch(value)) {
+      return 'Password must contain at least 1 digit.';
+    }
+    if (!hasSpecialCharRegEx.hasMatch(value)) {
+      return 'Password must contain at least 1 special character (@\$!%*?&).';
+    }
+    if (!hasMin8CharsRegEx.hasMatch(value)) {
+      return 'Password must be at least 8 characters long.';
+    }
+    // Valid password
```
[the remark this answers is not in the document]
Comment (nitzanpap) [reply lib/helper_functions/validations.dart]: @abenartz Done.

File: lib/helper_functions/validations.dart [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    return null;
+  }
+
+  // Returns null if the given string is a valid email, otherwise returns an informative message.
```
[the remark this answers is not in the document]
Comment (nitzanpap) [reply lib/helper_functions/validations.dart]: @abenartz Done.

File: lib/helper_functions/validations.dart [hunk 8]
```diff
@@ -0,0 +1,44 @@
+abstract class Validations {
+  // Returns null if the given string is a valid password, otherwise returns an informative message.
+  static String? passwordValidator(String? value) {
```
[the remark this answers is not in the document]
Comment (nitzanpap) [reply lib/helper_functions/validations.dart]: @abenartz Done.

File: lib/helper_functions/validations.dart [hunk 9]
```diff
@@ -0,0 +1,44 @@
+abstract class Validations {
+  // Returns null if the given string is a valid password, otherwise returns an informative message.
```
[the remark this answers is not in the document]
Comment (nitzanpap) [reply lib/helper_functions/validations.dart]: @abenartz Done.

File: lib/helper_functions/validations.dart [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    }
+    if (!hasMin8CharsRegEx.hasMatch(value)) {
+      return 'Password must be at least 8 characters long.';
+    }
+    // Valid password
+    return null;
+  }
+
+  // Returns null if the given string is a valid email, otherwise returns an informative message.
+  static String? emailValidator(String? value) {
```
[the remark this answers is not in the document]
Comment (nitzanpap) [reply lib/helper_functions/validations.dart]: @abenartz Done.

File: lib/screens/signup_login_screen.dart [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+class SignupLoginScreen extends StatelessWidget {
+  final bool isLogin;
+
+  const SignupLoginScreen({super.key, this.isLogin = false});
+  static const int logoSectionSize = 40;
+
+  @override
+  Widget build(BuildContext context) {
+    const logoLayered = LogoLayered();
```
[the remark this answers is not in the document]
Comment (nitzanpap) [reply lib/screens/signup_login_screen.dart]: @abenartz Done.

**Review by abenartz — APPROVED** [review]

**Pull request merged**